### PR TITLE
Send pulse triggers should respect report timezone

### PR DESCRIPTION
**Pull request by qnkhuat**
We use the cron job to send pulse triggers, but we need to ensure it's triggered in the `report-timezone`; this PR ensures we do that and also makes changing the report-timezone update all trigger's timezone.

Reported in: https://metaboat.slack.com/archives/CKZEMT1MJ/p1715331817582219

### CODE REVIEW

File: src/metabase/task.clj [hunk 1]
```diff
@@ -265,7 +265,7 @@
    :priority           (.getPriority trigger)
    :start-time         (.getStartTime trigger)
    :may-fire-again?    (.mayFireAgain trigger)
-   :data               (.getJobDataMap trigger)})
+   :data               (into {} (.getJobDataMap trigger))})
```
Comment (calherries) [inline src/metabase/task.clj:268]: Why did you need to do this?

Comment (qnkhuat) [reply src/metabase/task.clj:268]: we need this in tests because we can't compare JobDataMap with clojure map

**Comment by calherries** [comment]
Why should updating the report time zone change the trigger of every pulse? For example, if I schedule an alert at midnight when the report time zone is time zone A, then I change the report time zone to time zone B, I would still expect the alert to fire at midnight in time zone A. Maybe that's how it worked previously, but that seems very surprising.

File: src/metabase/task/send_pulses.clj [hunk 2]
```diff
@@ -56,16 +60,24 @@
     (catch Throwable e
       (log/errorf e "Error sending Pulse %d to channel ids: %s" pulse-id (str/join ", " channel-ids)))))
 
+(defn- send-trigger-timezone
+  []
+  (or (driver/report-timezone)
+      (qp.timezone/system-timezone-id)
+      "UTC"))
```
Comment (calherries) [inline src/metabase/task/send_pulses.clj:67]: This feels like it could be DRY'd up with other parts of the codebase outside of pulses. It doesn't feel right to have a notion of the trigger time zone specific to send pulses task.

Also, before we were just using `(driver/report-timezone)` without `(qp.timezone/system-timezone-id)`, can you explain the change?

**Comment by qnkhuat** [comment]
> Why should updating the report time zone change the trigger of every pulse? 
answered in https://metaboat.slack.com/archives/C0641E4PB9B/p1715349254844079

File: src/metabase/task/send_pulses.clj (hunk 2 above)
Comment (qnkhuat) [reply src/metabase/task/send_pulses.clj:67]: it's the fallback timezone id we used in QP so I think we should apply it for pulse too: https://github.com/metabase/metabase/blob/907da2dd943508e22c9fc2db63fee066df6b270d/src/metabase/query_processor/timezone.clj#L90

It's also used in other tasks like persist refresh : https://github.com/metabase/metabase/blob/send-pulse-trigger-respect-timezone/src/metabase/task/persist_refresh.clj#L286

**Review by crisptrutski — COMMENTED** [review]
LGTM but will let Cal have final say

File: src/metabase/driver.clj [hunk 3]
```diff
@@ -51,6 +51,11 @@
     timezone-id
     (str (t/zone-id))))
 
+(defn- update-send-pulse-triggers-timezone!
+  []
+  (classloader/require 'metabase.task.send-pulses)
+  ((resolve 'metabase.task.send-pulses/update-send-pulse-triggers-timezone!)))
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/driver.clj:57]: yes, it's one of my annoyance as well!

File: src/metabase/task/send_pulses.clj (hunk 2 above)
Comment (calherries) [reply src/metabase/task/send_pulses.clj:67]: Right, I guess we should DRY this up. Instead of duplicating this logic again, `update-send-pulse-triggers-timezone!`, and the `cron-schedule` function should use the same function. Why not use `results-timezone-id` from the QP code?

Comment (qnkhuat) [reply src/metabase/task/send_pulses.clj:67]: `results-timezone-id` requires a DB. In case of pulse a dashboard can contains questions from more than 1 DB.

Wdyt if we move this to `metabase.task` or a new `metabase.task.util`?

Comment (calherries) [reply src/metabase/task/send_pulses.clj:67]: I don't mind where it goes too much. But I'm seeing a couple more places this could be DRYed up too:

https://github.com/metabase/metabase/blob/e6f47956c5a22c8052ae2db201f087cd6dcc4eac/src/metabase/email/messages.clj#L577-L579

https://github.com/metabase/metabase/blob/e6f47956c5a22c8052ae2db201f087cd6dcc4eac/src/metabase/task/index_values.clj#L82-L84

**Review by calherries — APPROVED** [review]
Works for me as long as the duplication of logic around the reporting time zone is resolved. That could be left to another PR though, because it looks like an existing issue
